src/backend/blt.rs

```diff
@@ -0,0 +1,284 @@
+use self::bear_lib_terminal::terminal::{self, Event as BltEvent, KeyCode};
+use std::collections::BTreeMap;
+use std::io::Write;
+use ::theme::{BaseColor, Color, ColorStyle, Effect};
+
+pub struct Concrete {
+    colours: BTreeMap<i16, (BltColor, BltColor)>,
+}
+
+impl backend::Backend for Concrete {
+    fn init() -> Self {
+        // TODO: Make this configurable?
+        terminal::open("Cursive", 80, 80);
+
+        Concrete { colours: BTreeMap::new() }
+    }
+
+    fn finish(&mut self) {
+        terminal::close();
+    }
+
+    fn init_color_style(&mut self, style: ColorStyle, foreground: &Color,
+                        background: &Color) {
+        self.colours.insert(style.id(),
+                            (colour_to_blt_colour(foreground),
+                             colour_to_blt_colour(background)));
+    }
+
+    fn with_color<F: FnOnce()>(&self, color: ColorStyle, f: F) {
+        let (fg, bg) = self.colours[&color.id()];
+        terminal::with_colors(fg, bg, f);
+    }
+
+    fn with_effect<F: FnOnce()>(&self, effect: Effect, f: F) {
+        match effect {
+            Effect::Simple => f(),
+            // TODO: how to do this correctly?`
+            //       BLT itself doesn't do this kind of thing,
+            //       we'd need the colours in our position,
+            //       but `f()` can do whatever
+            Effect::Reverse => {
+                terminal::with_colors(BltColor::from_rgb(0, 0, 0),
+                                      BltColor::from_rgb(255, 255, 255),
+                                      f)
+            }
+        }
+    }
+
+    fn has_colors(&self) -> bool {
+        true
+    }
+
+    fn screen_size(&self) -> (usize, usize) {
+        let Size { width, height } = terminal::state::size();
+        (width as usize, height as usize)
+    }
+
+    fn clear(&self) {
+        terminal::clear(None);
+    }
+
+    fn refresh(&mut self) {
+        terminal::refresh();
+    }
+
+    fn print_at(&self, (x, y): (usize, usize), text: &str) {
+        terminal::print_xy(x as i32, y as i32, text);
+    }
+
+    fn set_refresh_rate(&mut self, fps: u32) {
+        // TODO: unsupported
+    }
+
+    fn poll_event(&self) -> Event {
+        if let Some(ev) = terminal::wait_event() {
+            match ev {
+                // TODO: what should we do here?
+                BltEvent::Close => Event::Refresh,
+                BltEvent::Resize { .. } => Event::WindowResize,
+                // TODO: mouse support
+                BltEvent::MouseMove { .. } => Event::Refresh,
+                BltEvent::MouseScroll { .. } => Event::Refresh,
+                BltEvent::KeyPressed { key, ctrl, shift } => {
+                    blt_keycode_to_ev(key, shift, ctrl)
+                }
+                // TODO: there's no Key::Shift/Ctrl for w/e reason
+                BltEvent::ShiftPressed => Event::Refresh,
+                BltEvent::ControlPressed => Event::Refresh,
```

Comment:
I didn't get Shift and Control to send a message to the application, the terminal emulator was usually buffering it.
It may have been a ncurses limitation, do you ever get those events?